examples/src/main/java/org/reactivestreams/example/unicast/AsyncStreamPublisher.java

```diff
@@ -0,0 +1,288 @@
+    final class SubscriptionImpl implements Subscription, Runnable {
+        final Subscriber<? super T> subscriber; // We need a reference to the `Subscriber` so we can talk to it
+        private boolean cancelled = false; // This flag will track whether this `Subscription` is to be considered cancelled or not
+        private long demand = 0; // Here we track the current demand, i.e. what has been requested but not yet delivered
+        private T nextElementToBeSent = null; // we need to fetch an element by the subscription initialization to ensure the stream is not empty
+
+        SubscriptionImpl(final Subscriber<? super T> subscriber) {
+            // As per rule 1.09, we need to throw a `java.lang.NullPointerException` if the `Subscriber` is `null`
+            if (subscriber == null) {
+                throw null;
+            }
+            this.subscriber = subscriber;
+        }
+
+        // This `ConcurrentLinkedQueue` will track signals that are sent to this `Subscription`, like `request` and `cancel`
+        private final ConcurrentLinkedQueue<Signal> inboundSignals = new ConcurrentLinkedQueue<Signal>();
+
+        // We are using this `AtomicBoolean` to make sure that this `Subscription` doesn't run concurrently with itself,
+        // which would violate rule 1.3 among others (no concurrent notifications).
+        private final AtomicBoolean on = new AtomicBoolean(false);
+
+        // This method will register inbound demand from our `Subscriber` and validate it against rule 3.9 and rule 3.17
+        private void doRequest(final long n) {
+            if (n < 1) {
+                terminateDueTo(new IllegalArgumentException(subscriber + " violated the Reactive Streams rule 3.9 by requesting a non-positive number of elements."));
+            } else if (demand + n < 1) {
+                // As governed by rule 3.17, when demand overflows `Long.MAX_VALUE` we treat the signalled demand as "effectively unbounded"
+                demand = Long.MAX_VALUE;  // Here we protect from the overflow and treat it as "effectively unbounded"
+                doSend(); // Then we proceed with sending data downstream
+            } else {
+                demand += n; // Here we record the downstream demand
+                doSend(); // Then we can proceed with sending data downstream
+            }
+        }
+
+        // This handles cancellation requests, and is idempotent, thread-safe and not synchronously performing heavy computations as specified in rule 3.5
+        private void doCancel() {
+            cancelled = true;
+        }
+
+        // Instead of executing `subscriber.onSubscribe` synchronously from within `Publisher.subscribe`
+        // we execute it asynchronously, this is to avoid executing the user code (`Iterable.iterator`) on the calling thread.
+        // It also makes it easier to follow rule 1.9
+        private void doSubscribe() {
+            if (!cancelled) {
```

Comment:
`cancelled` being true here is impossible because the only way to get that true is by submitting a `Subscription` to the `Subscriber` that happens once and under this condition.